mono/metadata/threadpool.c

```diff
@@ -320,12 +326,15 @@ worker_callback (void)
 	 */
 	mono_defaults.threadpool_perform_wait_callback_method->save_lmf = TRUE;
 
+	gsize name_generation = thread->name.generation;
+	/* Set the name if this is the first call to worker_callback on this thread */
+	if (name_generation == 0)
+	   name_generation = set_thread_name (thread);
```

Comment:
This shouldn't be needed. Unless you want it set outside the loop the first time. The tilde should make the first time through the loop set it, and each time through the loop if there was an intervening change.

You might as well change this to just:
`name_generation = set_thread_name` w/o the `if`.

Generation should also be more opaque. It could start at `-1` if that was useful.

Reply:
The tilde is supposed to force it to run the first time through the loop. Perhaps that is too obscure? Or broken?

Reply:
I'll step through it later today, make sure it runs once and only once per thread in master.